File: automovil_g31/clases/Automovil.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace automovil_G3.clases
{
    class Automovil
    {
        public enum l_colores { Negro, Blanco, Gris, Rojo, Azul };
        //atributos
        private string marca;
        private l_colores color;
        private byte nro_puertas;
        private ushort max_velocidad;
        private ushort cilindraje;
        private byte nro_pasajeros;
        private float largo;
        private string modelo;
        private ushort ano;
        private float peso;
        private byte capacidad_tanque;
        //solamente los tocaran los metodos en la clase
        //EL USUARIO SOLO PODRA VERNOS
        private float nivel_tanque;
        private bool encendido = false;

        //accesores
        public string Marca
        {
            get
            {
                return marca.ToUpper();//devuelve el atributo marca
            }
            set//valido antes de escribir el atributo
            {
                if (string.IsNullOrEmpty(value) || string.IsNullOrWhiteSpace(value) || value.Length < 3)
                    throw new Exception("Revise la marca que no sea nula y que tenga más de 3 caracteres");
                else
                    marca = value;//al atributo marca le llevo el valor
            }
        }

        public l_colores Color
        {
            get
            {
                return color;//devuelve el atributo marca
            }
            set//valido antes de escribir el atributo
            {
                color = value;
            }
        }
        public byte Nro_puertas
        {
            get
            {
                return nro_puertas;//devuelve el atributo marca
            }
            set//valido antes de escribir el atributo
            {
                if (value < 1 || value > 6)
                    throw new Exception("Revise el número de puertas, debe estar entre 1 y 6");
                else
                    nro_puertas = value;
            }
        }
        public ushort Max_velocidad
        {
            get
            {
                return max_velocidad;
            }
            set
            {
                if (value > 532 || value < 120)
                    throw new Exception("Revise la velocidad máxima, entre 120 y 532");
                else
                    max_velocidad = value;
            }
        }
        public ushort Cilindraje
        {
            get
            {
                return cilindraje;
            }
            set
            {
                if (value < 800 || value > 7000)
                    throw new Exception("Revise el cilindraje, entre 800 y 7000cc");
                else
                    cilindraje = value;
            }
        }
        public byte Nro_pasajeros
        {
            get
            {
                return nro_pasajeros;
            }
            set
            {
                if (value < 1 || value > 7)
                    throw new Exception("Revise el núemro de pasajeros, entre 1 y 7");
                else
                    nro_pasajeros = value;
            }
        }
        public float Largo
        {
            get
            {
                return largo;
            }
            set
            {
                if (value < 1.5f || value > 5)
                    throw new Exception("Revise el largo, entre 1.5 y 5 metros");
                else
                    largo = value;
            }
        }
        public string Modelo
        {
            get
            {
                return modelo;
            }
            set
            {
                if (string.IsNullOrEmpty(value) || string.IsNullOrWhiteSpace(value) || modelo.Length < 1)
                    throw new Exception("Revise el modelo que no sea nula y que tenga mínimo 1 caracter");
                else
                    modelo = value;
            }

        }
        public ushort Ano
        {
            get
            {
                return ano;
            }
            set
            {
                if (value < 2000 || value > DateTime.Now.Year + 1)
                    throw new Exception("Revise el año, no debe ser menor a 2000 ni mayor a " + (DateTime.Now.Year + 1));
                else
                    ano = value;
            }
        }
        public float Peso
        {

            get
            {​​​​
              return peso;
            }​​​​
              set
            {​​​​
                 if (value < 690 || value > 5000)
                    throw new Exception("Revise el peso, no puede ser mayor que 5000kl ni menor que 690kl");
                else
                    peso = value;
            }​​​​



}​​​​
            public byte Capacidad_tanque
        {​​​​
            get
            {​​​​
             return capacidad_tanque;
            }​​​​
              set
            {​​​​
               if (value < 5 || value > 20)
                    throw new Exception("Revise la capacidad del tanque, no puede ser mayor que 20gl ni menor que 5 gl");
                else
                    capacidad_tanque = value;
            }​​​​
}​​​​
      //SOLO PUEDEN VER EL VALOR
       public float Nivel_tanque { get => nivel_tanque; }
        public bool Encendido { get => encendido; }

        public Automovil()//constructor con parametros
        {

        }
        public Automovil(string marca, l_colores color, byte nro_puertas, ushort max_velocidad,
       ushort cilindraje, byte nro_pasajeros, float largo, string modelo, ushort ano,
       float peso, byte capacidad_tanque)
        {
            Marca = marca;//el accesor recibe el parametro para validarlo
            Nro_pasajeros = nro_pasajeros;
            Color = color;
            Nro_puertas = nro_puertas;
            Max_velocidad = max_velocidad;
            Cilindraje = cilindraje;
            Largo = largo;
            Modelo = modelo;
            Ano = ano;
            Capacidad_tanque = capacidad_tanque;
            Peso = peso;
        }
        //Metodos

        public void Encender()
        {
            //enciende un vehiculo cuando esta apagado
            try
            {
                if (!encendido && nivel_tanque > 0) encendido = true;
                else
                    throw new Exception("no se puede encender por que no tiene" + "combustible");
            }
            catch (Exception)
            {
                throw new Exception("error en el metodo encender");
            }
        }
        public void Apagar()
        {
            //Apago el auto cuando esta encendido
            // (encendido) encendido equivale (encendido== true)
            // (!encendido) encendido equivale (encendido== false)


            try
            {
                if (encendido)
                    encendido = false;
                else
                    throw new Exception("el auto esta apagado");
            }
            catch (Exception)
            {
                throw new Exception("error en el metodo apagar");
            }
        }
        public void Cargar_Combustible()//llenar el tanque
                                        //1.verificar carro apagago
                                        //2. capacidad de tanque-nivel de combustible
        {
            try
            {
                if (!encendido && nivel_tanque < capacidad_tanque)
                    nivel_tanque += capacidad_tanque - nivel_tanque;
                else throw new Exception("debe de apagar el auto o el tanque esta lleno");
            }
            catch (Exception)
            {
                throw new Exception("cargar combustible");
            }
        }
        public bool Cargar_Combustible(uint Cant_dinero, uint precio_galon)
        {
            //1 auto apagado
            // 2 cant dinero y precio galon
            //3 convertir dinero en 
            // 4 nivel del tanque - nivel de combustible < galones ingresar
            //5 aumento nivel tanque en cantidad de galones que compre
            try
            {
                if (!encendido && Cant_dinero > 0 && precio_galon > 0)
                {
                    float cantidad_galones = (Cant_dinero / precio_galon);
                    if ((cantidad_galones + nivel_tanque) > capacidad_tanque)
                        throw new Exception("la cantidad de galones que estoy comprando excede" + "la capacidad del tanque");
                    else
                    {
                        nivel_tanque += cantidad_galones;
                        return true;

                    }
                }
                else
                    throw new Exception(" revisar la cantidad del dinero y el auto debe de estar apagado");
            }
            catch (Exception)
            {
                throw new Exception("error en el metodo encender");
            }
        }
        public bool Cargar_Combustible(float cantidad_galones)
        {
            //1 auto apagado
            //2 cantidad de galones a cargar sea 
            try
            {
                if (cantidad_galones > 0 && !encendido && (cantidad_galones + nivel_tanque) <= capacidad_tanque)
                {
                    nivel_tanque += cantidad_galones;
                    return true;
                }
                else
                    throw new Exception("revisar la cantidad de galones a cargar si excede la capacidad del tanque" + "o la cantidad de galones debe ser mayores a 0 y " +
                        "el auto debe de estar apagado");
            }
            catch (Exception)
            {
                throw new Exception("cargar combustible,por galones");
            }
        }

    }
}

File: automovil_g31/clases/Automovil.py
class Automovil:
    class Colores:
        NEGRO = "Negro"
        BLANCO = "Blanco"
        GRIS = "Gris"
        ROJO = "Rojo"
        AZUL = "Azul"

    def __init__(self, marca=None, color=None, nro_puertas=None, max_velocidad=None,
                 cilindraje=None, nro_pasajeros=None, largo=None, modelo=None,
                 ano=None, peso=None, capacidad_tanque=None):
        self._marca = None
        self._color = None
        self._nro_puertas = None
        self._max_velocidad = None
        self._cilindraje = None
        self._nro_pasajeros = None
        self._largo = None
        self._modelo = None
        self._ano = None
        self._peso = None
        self._capacidad_tanque = None
        self._nivel_tanque = 0
        self._encendido = False

        if marca:
            self.marca = marca
        if color:
            self.color = color
        if nro_puertas:
            self.nro_puertas = nro_puertas
        if max_velocidad:
            self.max_velocidad = max_velocidad
        if cilindraje:
            self.cilindraje = cilindraje
        if nro_pasajeros:
            self.nro_pasajeros = nro_pasajeros
        if largo:
            self.largo = largo
        if modelo:
            self.modelo = modelo
        if ano:
            self.ano = ano
        if peso:
            self.peso = peso
        if capacidad_tanque:
            self.capacidad_tanque = capacidad_tanque

    @property
    def marca(self):
        return self._marca.upper()

    @marca.setter
    def marca(self, value):
        if not value or len(value) < 3:
            raise ValueError("Revise la marca que no sea nula y que tenga más de 3 caracteres")
        self._marca = value

    @property
    def color(self):
        return self._color

    @color.setter
    def color(self, value):
        self._color = value

    @property
    def nro_puertas(self):
        return self._nro_puertas

    @nro_puertas.setter
    def nro_puertas(self, value):
        if value < 1 or value > 6:
            raise ValueError("Revise el número de puertas, debe estar entre 1 y 6")
        self._nro_puertas = value

    @property
    def max_velocidad(self):
        return self._max_velocidad

    @max_velocidad.setter
    def max_velocidad(self, value):
        if value < 120 or value > 532:
            raise ValueError("Revise la velocidad máxima, entre 120 y 532")
        self._max_velocidad = value

    @property
    def cilindraje(self):
        return self._cilindraje

    @cilindraje.setter
    def cilindraje(self, value):
        if value < 800 or value > 7000:
            raise ValueError("Revise el cilindraje, entre 800 y 7000cc")
        self._cilindraje = value

    @property
    def nro_pasajeros(self):
        return self._nro_pasajeros

    @nro_pasajeros.setter
    def nro_pasajeros(self, value):
        if value < 1 or value > 7:
            raise ValueError("Revise el número de pasajeros, entre 1 y 7")
        self._nro_pasajeros = value

    @property
    def largo(self):
        return self._largo

    @largo.setter
    def largo(self, value):
        if value < 1.5 or value > 5:
            raise ValueError("Revise el largo, entre 1.5 y 5 metros")
        self._largo = value

    @property
    def modelo(self):
        return self._modelo

    @modelo.setter
    def modelo(self, value):
        if not value or len(value) < 1:
            raise ValueError("Revise el modelo que no sea nulo y que tenga mínimo 1 caracter")
        self._modelo = value

    @property
    def ano(self):
        return self._ano

    @ano.setter
    def ano(self, value):
        current_year = 2024
        if value < 2000 or value > current_year + 1:
            raise ValueError(f"Revise el año, no debe ser menor a 2000 ni mayor a {current_year + 1}")
        self._ano = value

    @property
    def peso(self):
        return self._peso

    @peso.setter
    def peso(self, value):
        if value < 690 or value > 5000:
            raise ValueError("Revise el peso, no puede ser mayor que 5000kg ni menor que 690kg")
        self._peso = value

    @property
    def capacidad_tanque(self):
        return self._capacidad_tanque

    @capacidad_tanque.setter
    def capacidad_tanque(self, value):
        if value < 5 or value > 20:
            raise ValueError("Revise la capacidad del tanque, no puede ser mayor que 20 galones ni menor que 5 galones")
        self._capacidad_tanque = value

    @property
    def nivel_tanque(self):
        return self._nivel_tanque

    @property
    def encendido(self):
        return self._encendido

    def encender(self):
        if not self._encendido and self._nivel_tanque > 0:
            self._encendido = True
        else:
            raise RuntimeError("No se puede encender porque no tiene combustible")

    def apagar(self):
        if self._encendido:
            self._encendido = False
        else:
            raise RuntimeError("El auto está apagado")

    def cargar_combustible(self):
        if not self._encendido and self._nivel_tanque < self._capacidad_tanque:
            self._nivel_tanque = self._capacidad_tanque
        else:
            raise RuntimeError("Debe apagar el auto o el tanque está lleno")

    def cargar_combustible_con_dinero(self, cant_dinero, precio_galon):
        if not self._encendido and cant_dinero > 0 and precio_galon > 0:
            cantidad_galones = cant_dinero / precio_galon
            if (cantidad_galones + self._nivel_tanque) > self._capacidad_tanque:
                raise RuntimeError("La cantidad de galones que está comprando excede la capacidad del tanque")
            self._nivel_tanque += cantidad_galones
            return True
        raise RuntimeError("Revise la cantidad del dinero y que el auto esté apagado")

    def cargar_combustible_con_galones(self, cantidad_galones):
        if cantidad_galones > 0 and not self._encendido and (cantidad_galones + self._nivel_tanque) <= self._capacidad_tanque:
            self._nivel_tanque += cantidad_galones
            return True
        raise RuntimeError("Revise la cantidad de galones a cargar si excede la capacidad del tanque o el auto debe estar apagado")
